DAOS-12638 tse: splice completion CBs list to reduce locking

bfaccini opened the pull request:
tse_task_complete_callback() must be called with associated (struct tse_sched_private *)->dsp_lock granted in order to prevent dtp_comp_cb_list link-list of struct tse_task_private being corrupted.

Required-githooks: true

### Before requesting gatekeeper:

* [ ] Two review approvals and any prior change requests have been resolved.
* [ ] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [ ] `Features:` (or `Test-tag*`) commit pragma was used or there is a reason documented that there are no appropriate tags for this PR.
* [ ] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [ ] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate watchers.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficient testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

Commit:
DAOS-12638 tse: call tse_task_complete_callback() with dsp_lock

tse_task_complete_callback() must be called with associated
(struct tse_sched_private *)->dsp_lock granted in order
to prevent dtp_comp_cb_list link-list of struct tse_task_private
being corrupted.

Required-githooks: true

Signed-off-by: Bruno Faccini <bruno.faccini@intel.com>

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

Commit:
DAOS-12368 tse: lock only to splice list

In order to avoid too much serialization of CBs
execution, only lock to splice list of CBs.

Required-githooks: true

Signed-off-by: Bruno Faccini <bruno.faccini@intel.com>

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 commented:
Test stage Unit Test on EL 8 completed with status UNSTABLE.  https://build.hpdd.intel.com/job/daos-stack/job/daos//view/change-requests/job/PR-12047/3/testReport/(root)/

daosbuild1 commented:
Test stage Unit Test with memcheck on EL 8 completed with status UNSTABLE.  https://build.hpdd.intel.com/job/daos-stack/job/daos//view/change-requests/job/PR-12047/3/testReport/(root)/

Commit:
DAOS-12368 tse: splice correct link-list

previous commit was wrongly splicing
dsp->dsp_complete_list instead of
dtp->dtp_comp_cb_list ...

Required-githooks: true

Signed-off-by: Bruno Faccini <bruno.faccini@intel.com>

Commit:
Merge branch 'master' into bfaccini/DAOS-12638

To hope fixing CI hung situations ...

Signed-off-by: Bruno Faccini <bruno.faccini@intel.com>

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

Commit:
DAOS-12638 tse: splice back completion CBs list if reinit

To avoid mising and leaking remaining completion CBs upon
task reinit, splice back the already spliced list into original
task's list before leaving tse_task_complete_callback() in this
case.

Required-githooks: true

Signed-off-by: Bruno Faccini <bruno.faccini@intel.com>

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.
